core/src/layers/madsim.rs

```diff
@@ -0,0 +1,358 @@
+        &self,
+        path: &str,
+        args: OpScan,
+    ) -> crate::Result<(RpScan, Self::BlockingPager)> {
+        panic!("blocking_scan is not supported in MadsimLayer");
+    }
+}
+
+pub struct MadsimReader {
+    data: Option<Bytes>,
+}
+
+impl oio::Read for MadsimReader {
+    fn poll_read(&mut self, _cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<crate::Result<usize>> {
+        if let Some(ref data) = self.data {
+            let len = data.len();
+            buf[..len].copy_from_slice(data);
+            Poll::Ready(Ok(len))
+        } else {
+            Poll::Ready(Ok(0))
+        }
+    }
+
+    fn poll_seek(&mut self, cx: &mut Context<'_>, pos: SeekFrom) -> Poll<crate::Result<u64>> {
+        todo!()
+    }
+
+    fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<crate::Result<Bytes>>> {
+        todo!()
+    }
+}
+
+impl oio::BlockingRead for MadsimReader {
+    fn read(&mut self, buf: &mut [u8]) -> crate::Result<usize> {
+        todo!()
+    }
+
+    fn seek(&mut self, pos: SeekFrom) -> crate::Result<u64> {
+        todo!()
+    }
+
+    fn next(&mut self) -> Option<crate::Result<Bytes>> {
+        todo!()
+    }
+}
+
+pub struct MadsimWriter {
+    path: String,
+    args: OpWrite,
+    sim_server_socket: SocketAddr,
+}
+
+impl oio::BlockingWrite for MadsimWriter {
+    fn write(&mut self, bs: Bytes) -> crate::Result<()> {
+        todo!()
+    }
+
+    fn append(&mut self, bs: Bytes) -> crate::Result<()> {
+        todo!()
+    }
+
+    fn close(&mut self) -> crate::Result<()> {
+        todo!()
+    }
+}
+
+#[async_trait]
+impl oio::Write for MadsimWriter {
+    async fn write(&mut self, bs: Bytes) -> crate::Result<()> {
+        let req = Request::Write(self.path.to_string(), bs);
+        let ep = Endpoint::connect(self.sim_server_socket)
+            .await
+            .expect("fail to connect to sim server");
+        let (tx, mut rx) = ep
+            .connect1(self.sim_server_socket)
+            .await
+            .expect("fail to connect1 to sim server");
+        tx.send(Box::new(req))
+            .await
+            .expect("fail to send request to sim server");
+        rx.recv()
+            .await
+            .expect("fail to recv response from sim server");
+        Ok(())
+    }
+
+    async fn append(&mut self, bs: Bytes) -> crate::Result<()> {
+        todo!()
+    }
+
+    async fn abort(&mut self) -> crate::Result<()> {
+        todo!()
+    }
+
+    async fn close(&mut self) -> crate::Result<()> {
+        Ok(())
+    }
+}
+
+pub struct MadsimPager {}
+
+#[async_trait]
+impl oio::Page for MadsimPager {
+    async fn next(&mut self) -> crate::Result<Option<Vec<Entry>>> {
+        todo!()
+    }
+}
+
+impl oio::BlockingPage for MadsimPager {
+    fn next(&mut self) -> crate::Result<Option<Vec<Entry>>> {
+        todo!()
+    }
+}
+
+/// A simulated server.
+#[derive(Default, Clone)]
+pub struct SimServer;
+
+impl SimServer {
+    pub async fn serve(addr: SocketAddr) -> Result<()> {
```

Comment:
After a re-visit of madsim's code, I found that it's not required to use `madsim::net` for simlulation. Instead, we can implement `Simulator` trait directly.

The `Simulator` is defined at https://github.com/madsim-rs/madsim/blob/main/madsim/src/sim/plugin.rs

And a real example is `FsSim`: https://github.com/madsim-rs/madsim/blob/main/madsim/src/sim/fs.rs

In `FsSim`, madsim use `Mutex<HashMap<NodeId, FsNodeHandle>>` for simlulation.

```rust
impl Simulator for FsSim {
    fn new(_rand: &GlobalRng, _time: &TimeHandle, _config: &Config) -> Self {
        Default::default()
    }

    fn create_node(&self, id: NodeId) {
        let mut handles = self.handles.lock();
        handles.insert(id, FsNodeHandle::new());
    }

    fn reset_node(&self, id: NodeId) {
        self.power_fail(id);
    }
}
```

I think we can borrow the idea here too. We can use a `Mutex<HashMap<String, (Metadata, Bytes)>>` for simulate a storage service. And implement `Simulator` for this.

In this way, we don't need to play with `SimServer` anymore.

Reply:
Good idea! I will try.